### PR TITLE
network: DHCP server Add support to transmit SIP server

### PR DESCRIPTION
**Pull request by ssahani**
Implements http://www.rfc-editor.org/rfc/rfc3361.txt

cc @yuwata 
```
Frame 134: 348 bytes on wire (2784 bits), 348 bytes captured (2784 bits) on interface 0
Ethernet II, Src: 42:65:85:d6:4e:32 (42:65:85:d6:4e:32), Dst: 1e:04:f8:b8:2f:d4 (1e:04:f8:b8:2f:d4)
Internet Protocol Version 4, Src: 192.168.5.1, Dst: 192.168.5.11
User Datagram Protocol, Src Port: 67, Dst Port: 68
Dynamic Host Configuration Protocol (ACK)
    Message type: Boot Reply (2)
    Hardware type: Ethernet (0x01)
    Hardware address length: 6
    Hops: 0
    Transaction ID: 0x7cc87cb4
    Seconds elapsed: 0
    Bootp flags: 0x0000 (Unicast)
    Client IP address: 0.0.0.0
    Your (client) IP address: 192.168.5.11
    Next server IP address: 0.0.0.0
    Relay agent IP address: 0.0.0.0
    Client MAC address: 1e:04:f8:b8:2f:d4 (1e:04:f8:b8:2f:d4)
    Client hardware address padding: 00000000000000000000
    Server host name not given
    Boot file name not given
    Magic cookie: DHCP
    Option: (53) DHCP Message Type (ACK)
        Length: 1
        DHCP: ACK (5)
    Option: (51) IP Address Lease Time
        Length: 4
        IP Address Lease Time: (3600s) 1 hour
    Option: (1) Subnet Mask (255.255.255.0)
        Length: 4
        Subnet Mask: 255.255.255.0
    Option: (3) Router
        Length: 4
        Router: 192.168.5.1
    Option: (6) Domain Name Server
        Length: 4
        Domain Name Server: 192.168.5.1
    Option: (42) Network Time Protocol Servers
        Length: 4
        Network Time Protocol Server: 192.168.1.1
    Option: (120) SIP Servers <=====here
        Length: 9
        SIP Server Encoding: IPv4 Address (1)
        SIP Server Address: 192.168.1.1
        SIP Server Address: 192.168.5.2
    Option: (101) TCode
        Length: 13
        TZ TCode: Europe/Berlin
    Option: (54) DHCP Server Identifier (192.168.5.1)
        Length: 4
        DHCP Server Identifier: 192.168.5.1
    Option: (255) End
        Option End: 255
```

```
cat /run/systemd/netif/state                                                                                                   ✔  ⚡  3148  16:40:51 
# This is private data. Do not parse.
OPER_STATE=routable
CARRIER_STATE=carrier
ADDRESS_STATE=routable
DNS=192.168.94.2 192.168.5.1
NTP=192.168.5.1
SIP=192.168.1.1 192.168.5.2

```

### CODE REVIEW

File: src/network/networkd-dhcp-common.c [hunk 1]
```diff
@@ -124,6 +156,7 @@ int config_parse_dhcp_use_ntp(
         return 0;
 }
 
+
```
Comment (yuwata) [inline src/network/networkd-dhcp-common.c]: Spurious new line.

File: src/network/networkd-link.c [hunk 2]
```diff
@@ -3597,6 +3597,23 @@ int link_save(Link *link) {
...
+                    link->dhcp_lease) {
+                        const struct in_addr *addresses;
+
+                        r = sd_dhcp_lease_get_sip(link->dhcp_lease, &addresses);
+                        if (r > 0)
+                                if (serialize_in_addrs(f, addresses, r, space, in4_addr_is_non_local) > 0)
+                                        space = true;
+                }
+
+
```
Comment (yuwata) [inline src/network/networkd-link.c]: Duplicated empty line.

File: src/network/networkd-link.h [hunk 3]
```diff
@@ -146,6 +146,7 @@ typedef struct Link {
         Set *dnssec_negative_trust_anchors;
 
         char **ntp;
+        char **sip;
```
Comment (yuwata) [inline src/network/networkd-link.h]: This seems never used.

File: src/network/networkd-link.c [hunk 4]
```diff
@@ -3597,6 +3597,23 @@ int link_save(Link *link) {
                                         space = true;
                 }
 
+                fputc('\n', f);
+
+                fputs("SIP=", f);
+                space = false;
+                fputstrv(f, link->sip ?: link->network->sip, NULL, &space);
```
Comment (yuwata) [inline src/network/networkd-link.c]: `link->sip` is always `NULL`.

File: src/network/networkd-dhcp-server.c [hunk 5]
```diff
@@ -345,3 +412,55 @@ int config_parse_dhcp_server_ntp(
...
+                        return 0;
+
+                r = in_addr_from_string(AF_INET, w, &a);
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, r,
+                                   "Failed to parse SIP server address '%s', ignoring: %m", w);
+                        continue;
+                }
+
+                m = reallocarray(n->dhcp_server_ntp, n->n_dhcp_server_sip + 1, sizeof(struct in_addr));
```
Comment (yuwata) [inline src/network/networkd-dhcp-server.c]: copy and paste error: `dhcp_server_ntp` -> `dhcp_server_sip`

File: src/systemd/sd-dhcp-server.h [hunk 6]
```diff
@@ -48,6 +48,7 @@ int sd_dhcp_server_configure_pool(sd_dhcp_server *server, struct in_addr *addres
 int sd_dhcp_server_set_timezone(sd_dhcp_server *server, const char *timezone);
 int sd_dhcp_server_set_dns(sd_dhcp_server *server, const struct in_addr ntp[], unsigned n);
 int sd_dhcp_server_set_ntp(sd_dhcp_server *server, const struct in_addr dns[], unsigned n);
```
Comment (yuwata) [inline src/systemd/sd-dhcp-server.h]: This is not your fault, but could you fix the name of arguments?

File: src/network/networkd-link.h (hunk 3 above)
Comment (ssahani) [reply src/network/networkd-link.h]: Yes will delete that. And api to access those sip servers

File: src/systemd/sd-dhcp-server.h (hunk 6 above)
Comment (ssahani) [reply src/systemd/sd-dhcp-server.h]: done

Comment (ssahani) [reply src/systemd/sd-dhcp-server.h]: you got hawk eyes for sure :)

**Comment by yuwata** [comment]
I modified the man page. And before that, all the CIs were green. Let's merge this.

**Comment by ssahani** [comment]
> I modified the man page. And before that, all the CIs were green. Let's merge this.

Thank you .

**Pull request merged**